src/java.base/share/classes/java/lang/annotation/ElementType.java

```diff
@@ -26,106 +26,214 @@
- * #FIELD}, {@link #LOCAL_VARIABLE}, {@link #METHOD}, {@link #PACKAGE}, {@link
- * #MODULE}, {@link #PARAMETER}, {@link #TYPE}, and {@link #TYPE_PARAMETER}
- * correspond to the declaration contexts in JLS {@jls 9.6.4.1}.
+ * {@snippet id='example' :
+ * @Target({ElementType.TYPE_PARAMETER, ElementType.LOCAL_VARIABLE})
+ * public @interface MyAnnotation {}
+ * }
  *
- * <p>For example, an annotation whose interface is meta-annotated with
- * {@code @Target(ElementType.FIELD)} may only be written as a modifier for a
- * field declaration.
+ * <h2 id="kinds-of-annotations">Declaration annotations and type-use
+ * annotations</h2>
  *
- * <p>The constant {@link #TYPE_USE} corresponds to the type contexts in JLS
- * {@jls 4.11}, as well as to two declaration contexts: class and interface
- * declarations (including annotation declarations) and type parameter
- * declarations.
+ * <p>Most annotations in Java code are <b>declaration
+ * annotations</b>, which act like modifiers of declarations (such as
+ * a field or method declaration). The constants of this class cover
+ * all ten kinds of annotatable declarations, plus a subcategory of
+ * {@link #TYPE} called {@link #ANNOTATION_TYPE}. An annotation
+ * interface can be used as a declaration annotation if it either
+ * omits {@link java.lang.annotation.Target @Target}, or uses it to
+ * list which specific kinds of declarations it should apply to.
  *
- * <p>For example, an annotation whose interface is meta-annotated with
- * {@code @Target(ElementType.TYPE_USE)} may be written on the class or
- * interface of a field (or within the class or interface of the field, if it
- * is a nested or parameterized class or interface, or array class), and may
- * also appear as a modifier for, say, a class declaration.
+ * <p>There are also <b>type-use annotations</b> (sometimes called
+ * "type annotations"), which can appear anywhere a Java type is being
+ * indicated (normally, immediately preceding that type). To be used
+ * as a type-use annotation, an annotation interface must use {@link
+ * java.lang.annotation.Target @Target} and include {@link #TYPE_USE}
+ * explicitly.
  *
- * <p>The {@code TYPE_USE} constant includes class and interface declarations
- * and type parameter declarations as a convenience for designers of
- * type checkers which give semantics to annotation interfaces. For example,
- * if the annotation interface {@code NonNull} is meta-annotated with
- * {@code @Target(ElementType.TYPE_USE)}, then {@code @NonNull}
- * {@code class C {...}} could be treated by a type checker as indicating that
- * all variables of class {@code C} are non-null, while still allowing
- * variables of other classes to be non-null or not non-null based on whether
- * {@code @NonNull} appears at the variable's declaration.
+ * <h3 id="ambiguous">Ambiguous contexts</h3>
+ *
+ * <p>For six kinds of declarations, type-use annotations can also
```

Comment:
The GitHub UI isn't letting me reply directly to the relevant existing comment. In any case, I strongly advise against referring to concrete numbers for language contexts in the JLS. As noted, that is a hazard for the comments getting out of date with language changes or previewing language changes. The JLS text does favor use of such numbers, but the JLS as a whole is in the same maintenance domain so there is a much higher chance that when adding another context, all the N's in the JLS get updated to (N+1)'s.